[Merged by Bors] - feat(Topology/KrullDimension): add subspace dimension inequality

a-dangelo opened the pull request:
This PR proves that subspaces have Krull dimension at most that of the ambient space: dim(Y) ≤ dim(X) for Y ⊆ X (theorem topologicalKrullDim_subspace_le).

Supporting results about IrreducibleCloseds were refactored and moved from KrullDimension.lean to Closeds.lean for better modularity.
Note: Some code/documentation generated with AI assistance (Gemini).

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

a-dangelo committed:
feat(Topology/KrullDimension): add subspace
  dimension inequality

  • Add mapIrreducibleClosed function for
  canonical maps between irreducible closed
  sets
  • Prove injectivity and strict monotonicity
  of these maps
  • Establish partial order structure on
  IrreducibleCloseds
  • Prove topKrullDim_subspace_le theorem
  showing dim(Y) ≤ dim(X) for subspaces

a-dangelo committed:
feat(Topology/KrullDimension): add subspace
  dimension inequality

  • Add mapIrreducibleClosed function for
  canonical maps between irreducible closed
  sets
  • Prove injectivity and strict monotonicity
  of these maps
  • Establish partial order structure on
  IrreducibleCloseds
  • Prove topKrullDim_subspace_le theorem
  showing dim(Y) ≤ dim(X) for subspaces

kim-em commented on Mathlib/Topology/KrullDimension.lean:
```diff
@@ -62,3 +74,123 @@ theorem IsHomeomorph.topologicalKrullDim_eq (f : X → Y) (h : IsHomeomorph f) :
+
+/-! ### Partial order structure on irreducible closed sets -/
+
+instance : PartialOrder (IrreducibleCloseds X) where
+  le s t := s.carrier ⊆ t.carrier
+  le_refl s := Set.Subset.refl _
+  le_trans s t u hst htu := Set.Subset.trans hst htu
+  le_antisymm s t hst hts := by
+    apply IrreducibleCloseds.ext
+    exact Set.Subset.antisymm hst hts
```
Again, all this stuff not actually about Krull dimension needs to be in earlier files.

ADA-Projects commented on Mathlib/Topology/KrullDimension.lean:
```diff
@@ -62,3 +74,123 @@ theorem IsHomeomorph.topologicalKrullDim_eq (f : X → Y) (h : IsHomeomorph f) :
+
+/-! ### Maps between irreducible closed sets -/
+
+/-- The canonical map from irreducible closed sets of a subspace `Y` to irreducible
+closed sets of the ambient space `X`, defined by taking the closure of the image
+under the inclusion map. This map is crucial for comparing Krull dimensions. -/
+def mapIrreducibleClosed (Y : Set X) (c : IrreducibleCloseds Y) : IrreducibleCloseds X where
+  carrier := closure (Subtype.val '' c.carrier)
+  is_irreducible' := c.is_irreducible'.image Subtype.val
+    (continuous_subtype_val.continuousOn) |>.closure
+  is_closed' := isClosed_closure
```
Completely agree, sorry for that. Where should I move them? in Irreducible.lean?

ADA-Projects commented on Mathlib/Topology/KrullDimension.lean:
```diff
@@ -62,3 +74,123 @@ theorem IsHomeomorph.topologicalKrullDim_eq (f : X → Y) (h : IsHomeomorph f) :
+    intro hx_in_A
+    change x ∈ B.carrier
+    rw [← B.is_closed'.closure_eq]
+    -- Use the mathlib lemma for embeddings
```
Yes, sorry I've used Gemini and I also refactored a bit without carefully checking everything and this slipped to me.

a-dangelo committed:
refactor(Topology): move IrreducibleCloseds extensions from KrullDimension to Sets/Closeds

Move PartialOrder instance, map functions, and embedding-related lemmas from
Topology/KrullDimension.lean to Topology/Sets/Closeds.lean for better code organization.
This separates basic IrreducibleCloseds operations from topological Krull dimension theory,
allowing other modules to use IrreducibleCloseds functionality without importing
dimension-specific concepts.

Changes:
- Move PartialOrder instance and le_iff_subset/lt_iff_subset lemmas to Sets/Closeds
- Move IrreducibleCloseds.map and related functions for closed continuous maps
- Move mapOfEmbedding functions and lemmas for general embeddings
- Move mapIrreducibleClosed functions for subspace embeddings
- Keep only topological Krull dimension definitions and theorems in KrullDimension
- Add necessary imports to Sets/Closeds for embedding support

a-dangelo committed:
fix: remove unused import Mathlib.Topology.Homeomorph.Lemmas

ADA-Projects commented:
-awaiting-author

a-dangelo committed:
Update Mathlib/Topology/KrullDimension.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

a-dangelo committed:
style(Topology): refactor IrreducibleCloseds.map and improve code organization

Address reviewer feedback with the following improvements:

- Simplify `IrreducibleCloseds.map` to work with any continuous function
  (no longer requires closed maps)
- Rename lemmas from `mapOfEmbedding_*` to `map_*_of_embedding` for
  consistency with Mathlib naming conventions
- Use `inferInstance` for PartialOrder instance on IrreducibleCloseds
- Merge variable declarations in Sets/Closeds.lean
- Add necessary imports (Topology.Defs.Induced)
- Open Topology namespace for cleaner code
- Simplify proofs and improve readability

The main theorem `topologicalKrullDim_subspace_le` remains unchanged.

ADA-Projects commented:
-awaiting-author

erdOne commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -430,6 +433,68 @@ order isomorphism. -/
+/-- The map `IrreducibleCloseds.map` is injective when `f` is an embedding.
+This relies on the property of embeddings that a closed set in the domain is the preimage
+of the closure of its image. -/
+lemma map_injective_of_embedding {f : β → α} (hf : Topology.IsEmbedding f) :
```
```suggestion
lemma map_injective_of_isEmbedding {f : β → α} (hf : Topology.IsEmbedding f) :
```

erdOne commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -430,6 +433,68 @@ order isomorphism. -/
+/-- The canonical map `mapIrreducibleClosed` is strictly monotone. -/
+lemma mapIrreducibleClosed_strictMono (β : Set α) :
+    StrictMono (mapIrreducibleClosed β) :=
+  map_strictMono_of_embedding Topology.IsEmbedding.subtypeVal
```
I am less sure if we want these. I think `map continuous_subtype_val` is good enough?

ADA-Projects replied:
yeah, I agree. I've removed them.

erdOne commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -430,6 +433,68 @@ order isomorphism. -/
+  carrier := closure (f '' c.carrier)
+  isIrreducible' := c.isIrreducible.image f hf.continuousOn |>.closure
+  isClosed' := isClosed_closure
+
```
Can you add a lemma saying `(map hf s : Set _) = closure (f '' s)`, which should be called `coe_map`? Thanks.

a-dangelo committed:
refactor(Topology): simplify IrreducibleCloseds.map and improve naming consistency

- Simplify readability in `IrreducibleCloseds.map` making the function explicit
- Rename `map_*_of_embedding` to `map_*_of_isEmbedding` for Mathlib consistency
- Add `coe_map` simp lemma for cleaner proofs
- Remove redundant `mapIrreducibleClosed` wrapper functions

erdOne commented:
I think you have messed up something. Can you try merging master again?

erdOne commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -430,6 +433,55 @@ order isomorphism. -/
 def orderIsoSubtype' : IrreducibleCloseds α ≃o { x : Set α // IsClosed x ∧ IsIrreducible x } :=
   equivSubtype'.toOrderIso (fun _ _ h ↦ h) (fun _ _ h ↦ h)
 
+/-! ### Partial order structure on irreducible closed sets and maps thereof-/
```
```suggestion
/-! ### Partial order structure on irreducible closed sets and maps thereof. -/
```

ADA-Projects commented on Mathlib/Topology/KrullDimension.lean:
```diff
@@ -26,32 +36,13 @@ noncomputable def topologicalKrullDim (T : Type*) [TopologicalSpace T] : WithBot
 theorem IsClosedEmbedding.topologicalKrullDim_le (f : X → Y) (hf : IsClosedEmbedding f) :
     topologicalKrullDim X ≤ topologicalKrullDim Y :=
-  krullDim_le_of_strictMono _ (IrreducibleCloseds.map_strictMono hf)
+  krullDim_le_of_strictMono _ (map_strictMono_of_isEmbedding hf.toIsEmbedding)
```
is the merge mess cleared? I made a mess because I've used the commands to commit simple changes from here, but then I forgot about it.

I will add the deprecated line.

erdOne commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -430,6 +433,55 @@ order isomorphism. -/
+    simp_rw [mem_preimage]
+    change f x ∈ closure (f '' A.carrier)
+    rw [h_closures_eq]
+    exact subset_closure (mem_image_of_mem f hx_in_B)⟩
```
This is true for `IsInducing` as well and here's a golf:
```
lemma map_injective_of_isEmbedding {f : β → α} (hf : Topology.IsInducing f) :
    Function.Injective (map f hf.continuous) := by
  intro A B h_images_eq
  apply SetLike.coe_injective
  replace h_images_eq : closure (f '' A) = closure (f '' B) := congr($h_images_eq)
  rw [← A.isClosed.closure_eq, hf.closure_eq_preimage_closure_image, h_images_eq, 
    ← hf.closure_eq_preimage_closure_image, B.isClosed.closure_eq]
```

ADA-Projects replied:
should the name become then *_of_isInducing?

And should I modify hypothesis everywhere from there on (also in KrullDim) to have "hf : Topology.IsInducing" or should I leave IsEmbedding in all the rest and try to fix that?

a-dangelo committed:
refactor: address reviewer feedback on IrreducibleCloseds

- Rename map_*_of_isEmbedding to map_*_of_isInducing for generality
- Simplify map_injective_of_isInducing proof using SetLike.coe_injective
- Simplify map_strictMono_of_isInducing using Monotone.strictMono_of_injective
- Use IsInducing instead of IsEmbedding in dimension theorems
- Add deprecation alias for IsClosedEmbedding.topologicalKrullDim_le
- Remove redundant PartialOrder instance (already inferred)

erdOne commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -430,6 +433,38 @@ order isomorphism. -/
+/-- The map `IrreducibleCloseds.map` is strictly monotone when `f` is an embedding. -/
+lemma map_strictMono_of_isInducing {f : β → α} (hf : Topology.IsInducing f) :
+    StrictMono (map f hf.continuous) := by
+  exact Monotone.strictMono_of_injective
```
```suggestion
    StrictMono (map f hf.continuous) :=
 Monotone.strictMono_of_injective
```

a-dangelo committed:
style: use term mode in map_strictMono_of_isInducing

erdOne commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -20,6 +22,7 @@ For a topological space `α`,
 
 
 open Order OrderDual Set
+open scoped Topology
```
I think you can just `open Topology` and remove the `Topology.` below.

erdOne commented on Mathlib/Topology/KrullDimension.lean:
```diff
@@ -26,39 +36,29 @@ noncomputable def topologicalKrullDim (T : Type*) [TopologicalSpace T] : WithBot
-lemma IrreducibleCloseds.map_strictMono {f : X → Y} (hf : IsClosedEmbedding f) :
-    StrictMono (IrreducibleCloseds.map hf.continuous hf.isClosedMap) :=
-  fun ⦃_ _⦄ UltV ↦ hf.injective.image_strictMono UltV
+/-- If `f : Y → X` is an embedding, then `dim(Y) ≤ dim(X)`.
```
```suggestion
/-- If `f : Y → X` is inducing, then `dim(Y) ≤ dim(X)`.
```

a-dangelo committed:
refactor: extract map_mono and simplify map_strictMono_of_isInducing.

erdOne reviewed (approved):
Thanks for the great PR and for the patience! 
maintainer delegate

a-dangelo committed:
refactor: removed unnecessary imports from Topology/Sets/Closeds

ADA-Projects commented on Mathlib/Topology/Sets/Closeds.lean:
```diff
@@ -5,6 +5,8 @@ Authors: Floris van Doorn, Yaël Dillies
 import Mathlib.Topology.Sets.Opens
 import Mathlib.Topology.Clopen
+import Mathlib.Topology.Constructions
+import Mathlib.Topology.Defs.Induced
```
Thanks for your patience as well!

riccardobrasca commented:
Can you please make the PR description shorter, like one or two sentences? There is no need to explain everything that happened in the review process (currently the description is longer than the PR itself...).
